### PR TITLE
Completely replace System.Web.Routing with implementation from reference source

### PR DESCRIPTION
**Pull request by pruiz**
Hi,

This pull-req completely replaces System.Web.Routing code using reference source's implementation instead.

Current mono's master already had a few parts of System.Web.Routing replaced by reference source's counterpart, but there were still a few classes missing, which were causing route address generation incompatibilities between mono & ms.net when running stacks like MVC.

### CODE REVIEW

**Comment by akoeplinger** [comment]
build

File: mcs/class/System.Web/System.Web.Security/FormsAuthenticationModule.cs [hunk 1]
```diff
@@ -71,6 +89,7 @@ public void Dispose ()
 
 		public void Init (HttpApplication app)
 		{
+
```
Comment (tritao) [inline mcs/class/System.Web/System.Web.Security/FormsAuthenticationModule.cs]: Extra new line here

Comment (akoeplinger) [inline mcs/class/System.Web/System.Web.Security/FormsAuthenticationModule.cs]: tbh, I couldn't care less about code style in this part of the codebase :laughing:

**Comment by akoeplinger** [comment]
@pruiz did you try running a website with these changes to make sure nothing important broke?

**Comment by pruiz** [comment]
I've just squashed this two commits into a single one and (force) pushed the branch again.

**Comment by pruiz** [comment]
@akoeplinger may you command monojenkins to build this again?

**Comment by akoeplinger** [comment]
build

**Comment by pruiz** [comment]
@akoeplinger Regarding testing, I have another version of this patch against mono 4.2.1.124 which has been running for a few days. And I did setup another one running against this version of the patch/framework yesterday. If it nothing weird happens in a couple of days I'll let you guys know.

**Comment by akoeplinger** [comment]
Given that all tests are passing and you said an MVC app didn't blow up immediately this is likely fine to merge :)

**Pull request merged**